pravega/pravega
# Issue 4260: No-Op tier-2 Adapter

Signed-off-by: kevinhan88 <kevinhan88@gmail.com>

**Change log description**  
. Created a thin No-Op (of payload) layer on top of SyncStorage, which doesn't change any existing functionality but skips the tier 2 payload writing (system segments still go through)
. Introduced two new config properties for the No-Op layer
. StorageLoader now detects the new config and load the new No-Op factory if necessary
. Created a new unit test to prove that existing functionalities still work, and the new No-Op layer works as expected.

**Purpose of the change**  
Closes #4260

**What the code does**  
(Detailed description of the code changes)
. Added NoOpStorage and NoOpStorageFactory - this is a layer on top of SyncStorage, which ignores non-system writing to tier-2
. Added default method createSyncStorage() into interface StorageFactory; this is to expose the syncStorage for No-Op factory to encapsulate.
. Changes HDFSStorageFactory, FileSystemStorageFactory, ExtendedS3StorageFactory and InMemoryStorageFactory to adopt the above interface change.
. Added storageextra.storageNoOpMode into config.properties with default value false. If this switch is set to true, NoOp mode will be turned on.
. Added storageextra.storageWriteNoOpLatencyMilli into config.properties with default value 20. When No-Op mode is turned on, this latency is applied to make the no-op operation real as possible.
. StorageExtraConfig created to accommodate the above two new config properties.
. StorageLoader is revised, when it loads storage implementation, it also checks storageextra.storageNoOpMode; if it's true, then the loader wraps the original implementation factory with the NoOp factory.
. StreamSegmentServiceNoOpStoreTests is added to test the No-Op layer. In order to verify no-operation, internally the test creates an internal in-memory storage to accommodate the segment which should be ignored. With this mechanism, the end-to-end test inside StreamSegmentStoreTestBase should pass as usually, without noticing any difference.

**How to verify it**  
(Optional: steps to verify that the changes are effective)
All existing tests, particularly those extend StreamSegmentStoreTestBase, should work without any change. 
The new unit test, StreamSegmentServiceNoOpStoreTests should pass.
